Core/AMPCanonicalExtractor.swift

```diff
@@ -0,0 +1,160 @@
+                                                                encodedContentRuleList: ruleSource) { [weak self] ruleList, error  in
+            guard error != nil else {
+                print(error?.localizedDescription ?? "AMPCanonicalExtractor - Error compiling image blocking rules")
+                return
+            }
+            
+            self?.imageBlockingRules = ruleList
+        }
+    }
+    
+    public func urlContainsAmpKeyword(_ url: URL?,
+                                      config: PrivacyConfiguration = PrivacyConfigurationManager.shared.privacyConfig) -> Bool {
+        linkCleaner.resetLastAmpUrl()
+        guard let url = url, !linkCleaner.isURLExcluded(url: url, config: config) else { return false }
+        let urlStr = url.absoluteString
+        
+        let ampKeywords = TrackingLinkSettings(fromConfig: config).ampKeywords
+        
+        for keyword in ampKeywords {
+            if urlStr.contains(keyword) {
+                return true
+            }
+        }
+        
+        return false
+    }
+    
+    private func buildUserScript() -> WKUserScript {
+        let source = """
+(function() {
+    document.addEventListener('DOMContentLoaded', (event) => {
+        const canonicalLinks = document.querySelectorAll('[rel="canonical"]')
+        window.webkit.messageHandlers.\(Constants.sendCanonical).postMessage({
+            \(Constants.canonicalKey): canonicalLinks.length > 0 ? canonicalLinks[0].href : undefined
+        })
+    })
+})()
+"""
+        
+        return WKUserScript(source: source, injectionTime: .atDocumentStart, forMainFrameOnly: true)
+    }
+    
+    public func getCanonicalUrl(initiator: URL?, url: URL?,
+                                config: PrivacyConfiguration = PrivacyConfigurationManager.shared.privacyConfig,
+                                completion: @escaping ((URL?) -> Void)) {
+        guard let url = url, !linkCleaner.isURLExcluded(url: url, config: config) else {
+            completion(nil)
+            return
+        }
+        
+        if let initiator = initiator, linkCleaner.isURLExcluded(url: initiator, config: config) {
+            completion(nil)
+            return
+        }
+        
+        self.completion = completion
+        
+        let configuration = WKWebViewConfiguration()
+        configuration.websiteDataStore = .nonPersistent()
+        configuration.userContentController.add(self, name: Constants.sendCanonical)
+        configuration.userContentController.addUserScript(buildUserScript())
+        if let rulesList = ContentBlockerRulesManager.shared.currentRules?.rulesList {
+            configuration.userContentController.add(rulesList)
+        }
+        if let imageBlockingRules = imageBlockingRules {
+            configuration.userContentController.add(imageBlockingRules)
+        }
+        
+        webView = WKWebView(frame: .zero, configuration: configuration)
+        webView?.load(URLRequest(url: url))
```

Comment:
What about completion handler in case connection times out and URL is not loaded?